performance_assessment.py

```diff
@@ -0,0 +1,189 @@
+        storage.add_image(image_path)
+
+    ocr = OCR(os.getenv("AZURE_API_ENDPOINT"), os.getenv("AZURE_API_KEY"))
+    gpt = GPT(
+        os.getenv("AZURE_OPENAI_ENDPOINT"),
+        os.getenv("AZURE_OPENAI_KEY"),
+        os.getenv("AZURE_OPENAI_DEPLOYMENT"),
+    )
+
+    # Run performance test
+    start_time = time.time()
+    actual_output = analyze(storage, ocr, gpt) # <-- the `analyse` function deletes the images it processes so we don't need to clean up our image copies
+    performance = time.time() - start_time
+
+    # Process actual output
+    actual_fields = extract_leaf_fields(json.loads(actual_output.model_dump_json()))
+
+    # Load expected output
+    with open(expected_json_path, 'r') as file:
+        expected_fields = extract_leaf_fields(json.load(file))
+
+    # Calculate accuracy
+    accuracy_results = calculate_accuracy(expected_fields, actual_fields)
+
+    # Return results
+    return {
+        'test_case_number': test_case_number,
+        'performance': performance,
+        'accuracy_results': accuracy_results,
+    }
+
+
+def generate_csv_report(results: list[dict[str, any]]) -> None:
+    timestamp = datetime.datetime.now().strftime("%Y%m%d%H%M")
+    os.makedirs("reports", exist_ok=True)
+    report_path = os.path.join("reports", f"test_results_{timestamp}.csv")
+
+    with open(report_path, mode='w', newline='') as file:
+        writer = csv.writer(file)
+        writer.writerow([
+            "Test Case",
+            "Field Name",
+            "Pass/Fail",
+            "Accuracy Score",
+            "Pipeline Speed (seconds)",
+            "Expected Value",
+            "Actual Value",
+        ])
+
+        for result in results:
+            test_case_number = result['test_case_number']
+            performance = result['performance']
+            for field_name, data in result['accuracy_results'].items():
+                writer.writerow([
+                    test_case_number,
+                    field_name,
+                    data['pass_fail'],
+                    f"{data['score']:.2f}",
+                    f"{performance:.4f}",
+                    data['expected_value'],
+                    data['actual_value'],
+                ])
+    print(f"CSV report generated and saved to: {report_path}")
+
+
+def main() -> None:
+    load_dotenv()
+
+    # Validate required environment variables
+    required_vars = [
+        "AZURE_API_ENDPOINT",
+        "AZURE_API_KEY",
+        "AZURE_OPENAI_ENDPOINT",
+        "AZURE_OPENAI_KEY",
+        "AZURE_OPENAI_DEPLOYMENT",
+    ]
+    missing_vars = [var for var in required_vars if not os.getenv(var)]
+    if missing_vars:
+        raise RuntimeError(f"Missing required environment variables: {', '.join(missing_vars)}")
+
+    test_cases = find_test_cases("test_data/labels")
+    results = []
+    for idx, (image_paths, expected_json_path) in enumerate(test_cases, 1):
+        result = run_test_case(idx, image_paths, expected_json_path)
```

Comment:
I suggest prints throughout the code to log checkpoints and for debugging. Consider the logging library for more control over the log levels.

Reply:
Yeah this is a good point. I just figured we could simply add proper logging and other QoL features later since this is suppose to be the barebone baseline implementation.

Reply:
I understand. Just simple prints should do for now. Because currently, especially with many test cases, the user has no idea if the script is even working.